server/controllers/searchController.ts

```diff
@@ -36,46 +35,64 @@ export default class SearchController {
     res.render('pages/search', viewModel)
   }
 
-  searchResultsPage: RequestHandler = async (req: Request, res: Response) => {
+  submitSearchQuery: RequestHandler = async (req: Request, res: Response) => {
     await this.auditService.logPageView(Page.SEARCH_RESULTS_PAGE, {
       who: res.locals.user.username,
       correlationId: req.id,
     })
 
     const validatedFormData: SearchOrderFormData = SearchOrderFormDataModel.parse(req.body)
 
-    // Validate input
     const validationErrors: ValidationResult = this.datastoreSearchService.validateInput({
       userToken: res.locals.user.token,
       data: validatedFormData,
     })
 
-    // If input validation fails, redirect to search view with errors
     if (validationErrors.length > 0) {
       req.session.formData = validatedFormData
       req.session.validationErrors = validationErrors
       res.redirect('search')
     } else {
-      // If input validation succeeds, execute the search
-      const results = await this.datastoreSearchService.search({
+      const queryExecutionId = await this.datastoreSearchService.submitSearchQuery({
         userToken: res.locals.user.token,
         data: validatedFormData,
       })
 
-      // Clear session data as it is no longer required
       req.session.validationErrors = undefined
       req.session.formData = undefined
 
-      await this.auditService.logPageView(Page.SEARCH_RESULTS_PAGE, {
-        who: res.locals.user.username,
-        correlationId: req.id,
-      })
+      // Redirect to results page
+      res.redirect(`search/results?search_id=${encodeURIComponent(queryExecutionId)}`)
+    }
+  }
 
-      // If results is Order[], proceed to results view
-      if (results.length > 0) {
-        res.render('pages/searchResults', { data: tabulateOrders(results as Order[]) })
+  searchResultsPage: RequestHandler = async (req: Request, res: Response) => {
```

Comment:
I feel this section is a bit messy, with some nested conditional logic.
Also I'm not sure if it's reasonable to handle routing using errors in this way, and in this location.

Reply:
A common JS pattern is to return after line 54, then you do not need the else statement.

You are not handling the model validation though?

Reply:
- Matt Smith provided refactoring advice to make the code more readable. This has been implemented.
- This controller no longer handles routing to separate 'no results' and 'results' pages. Instead the results page nunjucks template conditionally renders either a no results component, or a results component, based on the length of the results array.